chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
 
         Ok(iter)
     }
+
+    /// Task to perform catchup and creation of a flat storage shard spawned from a previous
+    /// resharding operation. This may be a long operation time-wise.
+    pub fn shard_catchup_task(
+        &self,
+        shard_uid: ShardUId,
+        flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> FlatStorageReshardingTaskStatus {
+        info!(target: "resharding", ?shard_uid, ?flat_head_block_hash, "flat storage shard catchup task started");
+        // If flat storage current status is beyond the chain final block, try again later.
+        let chain_final_head = chain_store.final_head().unwrap();
+        let height_result = chain_store.get_block_height(&flat_head_block_hash);
+        let Ok(height) = height_result else {
+            error!(target: "resharding", ?shard_uid, ?flat_head_block_hash, err = ?height_result.unwrap_err(), "can't find flat head height!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        if height > chain_final_head.height {
+            info!(target = "resharding", ?height, chain_final_height = ?chain_final_head.height, "flat head beyond chain final tip: postponing flat storage shard catchup task");
+            self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
+                resharder: self.clone(),
+                shard_uid,
+                flat_head_block_hash,
+            });
+            return FlatStorageReshardingTaskStatus::Cancelled;
+        }
+
+        // If the flat head is not in the canonical chain this task has failed.
+        match chain_store.get_block_hash_by_height(height) {
```

Comment:
That isn't what I expected the check for canonical to look like but alright :)

Reply:
Oh I don't know, it just surprised me. There is a function call `is_on_current_chain` in chain that seems to work in a similar way.